### ISSUE TITLE
Severe Performance Issue when Training/Evaluating CNNs from C#

### ISSUE THREAD

**Issue by nietras** [microsoft/CNTK#3753]
We are seeing a severe performance issue using CNTK from C# on Windows 10 x64. This issue exists in both CNTK 2.6.0 and 2.7.0 and we do not now the exact source of the problem.

We do detailed `Stopwatch` timings of our training code, which output something like the below.
```
|Train| #: 006336 Eval:    0.382271 Loss:     0.334825 Total 737.1 ms/b : Load  11.0 ms/b Train  18.3 ms/b GetEvalLoss 613.6 ms/b Dispose   0.7 ms/b
|Valid| #: 002487 Eval:    0.525908 Loss:     0.666885 Total 237.3 ms/b : Load  11.3 ms/b Eval. 224.0 ms/b GetEvalLoss   0.2 ms/b Dispose   0.8 ms/b
```
`Total` is the total time measured per minibatch `b` (minibatch size is 32 here), hence it takes a staggering 737.1 ms/b per minibatch for a rather simple wide resnet architecture for 3 channel images of width 192 x height 256.  This is on a 1080 Ti GPU, so DeviceDescriptor is default and hence GPU for all things defined. After total there are some detailed timings which are:

 * `Load` - time to load, augment (in CPU) input data and then manually create `Value`s for each channel input, a target vector and mask e.g. 8 measurements predicted.
 * `Train` - time to call `trainer.TrainMinibatch(inputVariableToValue, isSweepEnd, d);`
 * `GetEvalLoss` - time to call:
```
float evaluationValue = (float)trainer.PreviousMinibatchEvaluationAverage();
float lossValue = (float)trainer.PreviousMinibatchLossAverage();
```
 * `Dispose` - time to call `Value.Dispose();` for all input arguments.

For validation we call `TestMinibatch` to evaluate how model performs on validation data set after each data sweep/epoch. This is still very slow.

Now the timings says there is a massive cost calling `trainer.PreviousMinibatchEvaluationAverage()`, which should be close to zero. However, if you remove this, the cost or bottleneck simply moves e.g. it moves to `Dispose`. If we remove that, it moves to `Load`. Hence, there appears to be something that forces a post-train cost.

If profiling this you can see the perf problem occurs deep inside cuda BLAS e.g. here from Visual Studio.

![image](https://user-images.githubusercontent.com/10798831/65680753-e62d0f00-e057-11e9-9970-a3b904761131.png)

This stack trace is for CNTK 2.6.0, but the problem is the same for CNTK 2.7.0.

Note we see the same problem with regards to the perf bottleneck appearing to be the `PreviousMinibatchEvaluationAverage` for other problems and architectures e.g. simple AlexNets.

CC: @haixpham

**Comment by nietras** [comment]
Here is the same profile call tree for 2.7.0.

![image](https://user-images.githubusercontent.com/10798831/65681959-402ed400-e05a-11e9-8df4-b6102491c1a1.png)

Unfortunately, I cannot find symbols for either CNTK nor cublas to pinpoint methods.

**Comment by nietras** [comment]
Looking at this with NVidia NSight shows the following for CUDA Runtime API calls.

![image](https://user-images.githubusercontent.com/10798831/65682872-255d5f00-e05c-11e9-806e-2fd77ce8851e.png)

Hence, it appears perf is dominated by `cudaLaunchKernel` and `cudaMemcpy`

**Comment by nietras** [comment]
Function calls ordered by start time, and of duration > 10000us.

![image](https://user-images.githubusercontent.com/10798831/65683771-6ce4ea80-e05e-11e9-9c83-80526be818db.png)

**Comment by rhy-ama** [comment]
@nietras Any follow ups on this? Is this CUDA API /driver issue ?

**Comment by nietras** [comment]
@rhy-ama sadly no. Although, we did figure out that due to the way CNTK does evaluation/training **lazily** means that you can't rely on timing a given method in C# as being the source of bottleneck, it is simply then because the network has not done evaluating yet, even though execution has returned from a train or evaluate call.

**Comment by rhy-ama** [comment]
@nietras C# managed dll is just a wrapper to C++ code, that is heavily optimized. 

The easiest solution first. 32 batch size looks like 0.7 seconds per batch or 45 samples per second? I am i correct? This is not far off from what your GPU can handle if you use **double** precision. Moving your model to **float** ie single precision - or even **half** (if supported in C# - i do not recall). You should see something like 200 samples per second.

Second easiest is to install newer geforce driver for your card.

Now to troubleshooting.

How does C# training compare with using brainscript - to eliminate most of the issues re model size and batch size? 

That would give you a baseline. Just pick some standard model from CNTK samples so you do not have to re-write both C# and brainscipt (i recommend brainscript over python because it is easier to get performance read and separate things from CNTK and other factors).

When you do the evaluation via brainscript - on some test model of your choice - use something like msi afterburner (i assume you are doing all of this on windows with geforce drivers) to note the memory usage and GPU processor load.

In all, there are many different things that impact the speed of processing  

- size of the model and the batch size that fills the GPU and hence the speed of transfer from memory to GPU memory
- the hardware you have - PCI/PCIe bus lanes available for the GPU, the speed of of PCI/PCIe lanes, do you have NUMA enabled in bios, memory speed etc.
- the media where you have the data - data needs to be read from disk and fed to the model. if you are reading many small files from a network you will have huge latency. 

If, in your tests, you see GPU processor load oscillate between 0-100% and is not uniformly stay around 80%, then you have issues moving data to the GPU - either loading files (ideally locally from SSD) . If you are using network - look into other network topologies than TCP/IP (infiniband).
